genomic_data_service/rnaseq/domain/file.py

```diff
@@ -0,0 +1,174 @@
+from .expression import Expression
+import os
+
+from genomic_data_service.rnaseq.domain.expression import Expression
+from genomic_data_service.rnaseq.remote.tsv import local_tsv_iterable
+from genomic_data_service.rnaseq.remote.tsv import maybe_save_file
+from genomic_data_service.rnaseq.domain.constants import AT_ID
+from genomic_data_service.rnaseq.domain.constants import AT_TYPE
+from genomic_data_service.rnaseq.domain.constants import BASE_PATH
+from genomic_data_service.rnaseq.domain.constants import DATASET
+from genomic_data_service.rnaseq.domain.constants import DATASETS
+from genomic_data_service.rnaseq.domain.constants import DATASET_FIELDS
+from genomic_data_service.rnaseq.domain.constants import DOCUMENT_PREFIX
+from genomic_data_service.rnaseq.domain.constants import DOMAIN
+from genomic_data_service.rnaseq.domain.constants import EXPRESSION_AT_TYPE
+from genomic_data_service.rnaseq.domain.constants import EXPRESSION_ID
+from genomic_data_service.rnaseq.domain.constants import FILE_FIELDS
+from genomic_data_service.rnaseq.domain.constants import GENES
+from genomic_data_service.rnaseq.domain.constants import INDEXING_FIELDS
+
+
+ROW_VALUES = [
+    'gene_id',
+    'transcript_id(s)',
+    'TPM',
+    'FPKM',
+]
+
+
+def download_and_open_tsv(url, path):
+    maybe_save_file(url, path)
+    return local_tsv_iterable(path)
+
+
+def get_indices_from_header(header):
+    return [
+        header.index(value)
+        for value in ROW_VALUES
+    ]
+
+
+def get_tsv_header_and_indices(url, path):
+    tsv = download_and_open_tsv(url, path)
+    header = next(tsv)
+    indices = get_indices_from_header(header)
+    return tsv, header, indices
+
+
+def get_values_from_row(row, indices):
+    return [
+        row[index]
+        for index in indices
+    ]
+
+
+def get_expression_generator(url, path):
+    tsv, _, indices = get_tsv_header_and_indices(
+        url,
+        path
+    )
+    for row in tsv:
+        yield get_values_from_row(row, indices)
+
+
+def remove_version_from_gene_id(gene_id):
+    return gene_id.split('.')[0]
+
+
+class RnaSeqFile:
+
+    BASE_PATH = BASE_PATH
+    DOMAIN = DOMAIN
+    FILE_FIELDS = FILE_FIELDS
+    DATASET_FIELDS = DATASET_FIELDS
```

Comment:
we don't need to redefine an external constant in this context

Reply:
I do it like this in case you want to redefine quickly at runtime (e.g. in a test), like `File.DATASET_FIELDS = [...]`. Would be harder with only global constant.

Reply:
all good... we can always try mocking the constant